diff -ruN coreutils-8.27.orig/tests/fmt/base.pl coreutils-8.27/tests/fmt/base.pl
--- coreutils-8.27.orig/tests/fmt/base.pl	2017-01-01 23:34:24.000000000 +0100
+++ coreutils-8.27/tests/fmt/base.pl	2017-06-10 20:36:20.655861034 +0200
@@ -23,9 +23,6 @@
 
 my @Tests =
     (
-     ['8-bit-pfx', qw (-p 'ç'),
-      {IN=> "ça\nçb\n"},
-      {OUT=>"ça b\n"}],
      ['wide-1', '-w 32768',
       {ERR => "fmt: invalid width: '32768'\n"}, {EXIT => 1},
       {ERR_SUBST => $normalize_strerror}],
