--- coreutils-8.14/src/ls.c.org	2011-12-08 11:13:48.597770882 +0100
+++ coreutils-8.14/src/ls.c	2011-12-08 11:16:34.612089051 +0100
@@ -2583,6 +2583,11 @@
         }
       else
         break;
+
+        /* When processing a very large directory, and since we've inhibited
+           interrupts, this loop would take so long that ls would be annoyingly
+           uninterruptible.  This ensures that it handles signals promptly.  */
+        process_signals ();
     }
 
   if (closedir (dirp) != 0)
@@ -4041,9 +4046,9 @@
   if (stack)
     PUSH_CURRENT_DIRED_POS (stack);
 
+  process_signals ();
   if (used_color_this_time)
     {
-      process_signals ();
       prep_non_filename_text ();
       if (start_col / line_length != (start_col + width - 1) / line_length)
         put_indicator (&color_indicator[C_CLR_TO_EOL]);
