diff -ur coreutils-5.2.1/m4/fsusage.m4 coreutils-5.2.1.new/m4/fsusage.m4
--- coreutils-5.2.1/m4/fsusage.m4	2003-08-17 07:51:49.000000000 +0000
+++ coreutils-5.2.1.new/m4/fsusage.m4	2005-04-16 20:09:06.000000000 +0000
@@ -239,7 +239,7 @@
 # Prerequisites of lib/fsusage.c not done by jm_FILE_SYSTEM_USAGE.
 AC_DEFUN([gl_PREREQ_FSUSAGE_EXTRA],
 [
-  AC_REQUIRE([jm_AC_TYPE_UINTMAX_T])
+  AC_REQUIRE([gl_AC_TYPE_UINTMAX_T])
   AC_CHECK_HEADERS_ONCE(fcntl.h)
   AC_CHECK_HEADERS(dustat.h sys/fs/s5param.h sys/filsys.h sys/statfs.h sys/statvfs.h)
   jm_STATFS_TRUNCATES
diff -ur coreutils-5.2.1/m4/gettext.m4 coreutils-5.2.1.new/m4/gettext.m4
--- coreutils-5.2.1/m4/gettext.m4	2004-01-12 20:47:45.000000000 +0000
+++ coreutils-5.2.1.new/m4/gettext.m4	2005-04-16 20:09:06.000000000 +0000
@@ -358,19 +358,19 @@
   AC_REQUIRE([AC_C_INLINE])dnl
   AC_REQUIRE([AC_TYPE_OFF_T])dnl
   AC_REQUIRE([AC_TYPE_SIZE_T])dnl
-  AC_REQUIRE([jm_AC_TYPE_LONG_LONG])dnl
+  AC_REQUIRE([gl_AC_TYPE_LONG_LONG])dnl
   AC_REQUIRE([gt_TYPE_LONGDOUBLE])dnl
   AC_REQUIRE([gt_TYPE_WCHAR_T])dnl
   AC_REQUIRE([gt_TYPE_WINT_T])dnl
-  AC_REQUIRE([jm_AC_HEADER_INTTYPES_H])
-  AC_REQUIRE([jm_AC_HEADER_STDINT_H])
+  AC_REQUIRE([gl_AC_HEADER_INTTYPES_H])
+  AC_REQUIRE([gl_AC_HEADER_STDINT_H])
   AC_REQUIRE([gt_TYPE_INTMAX_T])
   AC_REQUIRE([gt_PRINTF_POSIX])
   AC_REQUIRE([AC_FUNC_ALLOCA])dnl
   AC_REQUIRE([AC_FUNC_MMAP])dnl
-  AC_REQUIRE([jm_GLIBC21])dnl
+  AC_REQUIRE([gl_GLIBC21])dnl
   AC_REQUIRE([gt_INTDIV0])dnl
-  AC_REQUIRE([jm_AC_TYPE_UINTMAX_T])dnl
+  AC_REQUIRE([gl_AC_TYPE_UINTMAX_T])dnl
   AC_REQUIRE([gt_HEADER_INTTYPES_H])dnl
   AC_REQUIRE([gt_INTTYPES_PRI])dnl
   AC_REQUIRE([gl_XSIZE])dnl
diff -ur coreutils-5.2.1/m4/glibc21.m4 coreutils-5.2.1.new/m4/glibc21.m4
--- coreutils-5.2.1/m4/glibc21.m4	2002-02-16 14:58:48.000000000 +0000
+++ coreutils-5.2.1.new/m4/glibc21.m4	2005-04-16 20:09:07.000000000 +0000
@@ -9,7 +9,7 @@
 # Test for the GNU C Library, version 2.1 or newer.
 # From Bruno Haible.
 
-AC_DEFUN([jm_GLIBC21],
+AC_DEFUN([gl_GLIBC21],
   [
     AC_CACHE_CHECK(whether we are using the GNU C Library 2.1 or newer,
       ac_cv_gnu_library_2_1,
diff -ur coreutils-5.2.1/m4/human.m4 coreutils-5.2.1.new/m4/human.m4
--- coreutils-5.2.1/m4/human.m4	2003-09-10 06:24:31.000000000 +0000
+++ coreutils-5.2.1.new/m4/human.m4	2005-04-16 20:09:07.000000000 +0000
@@ -10,7 +10,7 @@
 [
   dnl Prerequisites of lib/human.h.
   AC_REQUIRE([AM_STDBOOL_H])
-  AC_REQUIRE([jm_AC_TYPE_UINTMAX_T])
+  AC_REQUIRE([gl_AC_TYPE_UINTMAX_T])
 
   dnl Prerequisites of lib/human.c.
   AC_CHECK_HEADERS_ONCE(locale.h)
diff -ur coreutils-5.2.1/m4/intmax.m4 coreutils-5.2.1.new/m4/intmax.m4
--- coreutils-5.2.1/m4/intmax.m4	2004-01-12 20:47:45.000000000 +0000
+++ coreutils-5.2.1.new/m4/intmax.m4	2005-04-16 20:09:07.000000000 +0000
@@ -12,8 +12,8 @@
 
 AC_DEFUN([gt_TYPE_INTMAX_T],
 [
-  AC_REQUIRE([jm_AC_HEADER_INTTYPES_H])
-  AC_REQUIRE([jm_AC_HEADER_STDINT_H])
+  AC_REQUIRE([gl_AC_HEADER_INTTYPES_H])
+  AC_REQUIRE([gl_AC_HEADER_STDINT_H])
   AC_CACHE_CHECK(for intmax_t, gt_cv_c_intmax_t,
     [AC_TRY_COMPILE([
 #include <stddef.h> 
diff -ur coreutils-5.2.1/m4/intmax_t.m4 coreutils-5.2.1.new/m4/intmax_t.m4
--- coreutils-5.2.1/m4/intmax_t.m4	2003-08-17 07:51:49.000000000 +0000
+++ coreutils-5.2.1.new/m4/intmax_t.m4	2005-04-16 20:09:07.000000000 +0000
@@ -17,10 +17,10 @@
 [
   dnl For simplicity, we assume that a header file defines 'intmax_t' if and
   dnl only if it defines 'uintmax_t'.
-  AC_REQUIRE([jm_AC_HEADER_INTTYPES_H])
-  AC_REQUIRE([jm_AC_HEADER_STDINT_H])
+  AC_REQUIRE([gl_AC_HEADER_INTTYPES_H])
+  AC_REQUIRE([gl_AC_HEADER_STDINT_H])
   if test $jm_ac_cv_header_inttypes_h = no && test $jm_ac_cv_header_stdint_h = no; then
-    AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+    AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
     test $ac_cv_type_long_long = yes \
       && ac_type='long long' \
       || ac_type='long'
@@ -36,8 +36,8 @@
 
 AC_DEFUN([gt_AC_TYPE_INTMAX_T],
 [
-  AC_REQUIRE([jm_AC_HEADER_INTTYPES_H])
-  AC_REQUIRE([jm_AC_HEADER_STDINT_H])
+  AC_REQUIRE([gl_AC_HEADER_INTTYPES_H])
+  AC_REQUIRE([gl_AC_HEADER_STDINT_H])
   AC_CACHE_CHECK(for intmax_t, gt_cv_c_intmax_t,
     [AC_TRY_COMPILE([
 #include <stddef.h>
@@ -53,7 +53,7 @@
     AC_DEFINE(HAVE_INTMAX_T, 1,
       [Define if you have the 'intmax_t' type in <stdint.h> or <inttypes.h>.])
   else
-    AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+    AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
     test $ac_cv_type_long_long = yes \
       && ac_type='long long' \
       || ac_type='long'
diff -ur coreutils-5.2.1/m4/inttypes_h.m4 coreutils-5.2.1.new/m4/inttypes_h.m4
--- coreutils-5.2.1/m4/inttypes_h.m4	2003-07-10 08:35:02.000000000 +0000
+++ coreutils-5.2.1.new/m4/inttypes_h.m4	2005-04-16 20:09:07.000000000 +0000
@@ -11,7 +11,7 @@
 # Define HAVE_INTTYPES_H_WITH_UINTMAX if <inttypes.h> exists,
 # doesn't clash with <sys/types.h>, and declares uintmax_t.
 
-AC_DEFUN([jm_AC_HEADER_INTTYPES_H],
+AC_DEFUN([gl_AC_HEADER_INTTYPES_H],
 [
   AC_CACHE_CHECK([for inttypes.h], jm_ac_cv_header_inttypes_h,
   [AC_TRY_COMPILE(
diff -ur coreutils-5.2.1/m4/jm-macros.m4 coreutils-5.2.1.new/m4/jm-macros.m4
--- coreutils-5.2.1/m4/jm-macros.m4	2003-12-20 17:57:18.000000000 +0000
+++ coreutils-5.2.1.new/m4/jm-macros.m4	2005-04-16 20:09:07.000000000 +0000
@@ -130,7 +130,7 @@
     ])
 
   AM_LANGINFO_CODESET
-  jm_GLIBC21
+  gl_GLIBC21
   AM_ICONV
   jm_FUNC_UNLINK_BUSY_TEXT
 
@@ -249,8 +249,8 @@
   AC_CHECK_TYPE(major_t, unsigned int)
   AC_CHECK_TYPE(minor_t, unsigned int)
 
-  AC_REQUIRE([jm_AC_TYPE_UINTMAX_T])
-  AC_REQUIRE([jm_AC_TYPE_UNSIGNED_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_UINTMAX_T])
+  AC_REQUIRE([gl_AC_TYPE_UNSIGNED_LONG_LONG])
 
   AC_REQUIRE([AC_HEADER_MAJOR])
   AC_REQUIRE([AC_HEADER_DIRENT])
diff -ur coreutils-5.2.1/m4/longlong.m4 coreutils-5.2.1.new/m4/longlong.m4
--- coreutils-5.2.1/m4/longlong.m4	2003-08-17 07:51:49.000000000 +0000
+++ coreutils-5.2.1.new/m4/longlong.m4	2005-04-16 20:09:07.000000000 +0000
@@ -10,7 +10,7 @@
 
 # Define HAVE_LONG_LONG if 'long long' works.
 
-AC_DEFUN([jm_AC_TYPE_LONG_LONG],
+AC_DEFUN([gl_AC_TYPE_LONG_LONG],
 [
   AC_CACHE_CHECK([for long long], ac_cv_type_long_long,
   [AC_TRY_LINK([long long ll = 1LL; int i = 63;],
diff -ur coreutils-5.2.1/m4/mkstemp.m4 coreutils-5.2.1.new/m4/mkstemp.m4
--- coreutils-5.2.1/m4/mkstemp.m4	2003-09-13 10:45:30.000000000 +0000
+++ coreutils-5.2.1.new/m4/mkstemp.m4	2005-04-16 20:09:07.000000000 +0000
@@ -62,5 +62,5 @@
   AC_CHECK_HEADERS(stdint.h)
   AC_CHECK_FUNCS(__secure_getenv gettimeofday)
   AC_CHECK_DECLS_ONCE(getenv)
-  AC_REQUIRE([jm_AC_TYPE_UINTMAX_T])
+  AC_REQUIRE([gl_AC_TYPE_UINTMAX_T])
 ])
diff -ur coreutils-5.2.1/m4/prereq.m4 coreutils-5.2.1.new/m4/prereq.m4
--- coreutils-5.2.1/m4/prereq.m4	2003-12-18 10:33:33.000000000 +0000
+++ coreutils-5.2.1.new/m4/prereq.m4	2005-04-16 20:09:07.000000000 +0000
@@ -136,7 +136,7 @@
   AC_CHECK_HEADERS([netinet/in.h nfs/nfs_clnt.h nfs/vfs.h],,,
     [AC_INCLUDES_DEFAULT])
 
-  AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
 
   statxfs_includes="\
 $ac_includes_default
diff -ur coreutils-5.2.1/m4/stdint_h.m4 coreutils-5.2.1.new/m4/stdint_h.m4
--- coreutils-5.2.1/m4/stdint_h.m4	2003-07-10 08:34:01.000000000 +0000
+++ coreutils-5.2.1.new/m4/stdint_h.m4	2005-04-16 20:09:07.000000000 +0000
@@ -11,7 +11,7 @@
 # Define HAVE_STDINT_H_WITH_UINTMAX if <stdint.h> exists,
 # doesn't clash with <sys/types.h>, and declares uintmax_t.
 
-AC_DEFUN([jm_AC_HEADER_STDINT_H],
+AC_DEFUN([gl_AC_HEADER_STDINT_H],
 [
   AC_CACHE_CHECK([for stdint.h], jm_ac_cv_header_stdint_h,
   [AC_TRY_COMPILE(
diff -ur coreutils-5.2.1/m4/strtoimax.m4 coreutils-5.2.1.new/m4/strtoimax.m4
--- coreutils-5.2.1/m4/strtoimax.m4	2003-09-13 10:45:30.000000000 +0000
+++ coreutils-5.2.1.new/m4/strtoimax.m4	2005-04-16 20:09:07.000000000 +0000
@@ -33,5 +33,5 @@
 AC_DEFUN([gl_PREREQ_STRTOIMAX], [
   jm_AC_TYPE_INTMAX_T
   AC_CHECK_DECLS(strtoll)
-  AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
 ])
diff -ur coreutils-5.2.1/m4/strtoll.m4 coreutils-5.2.1.new/m4/strtoll.m4
--- coreutils-5.2.1/m4/strtoll.m4	2003-08-17 07:51:49.000000000 +0000
+++ coreutils-5.2.1.new/m4/strtoll.m4	2005-04-16 20:09:07.000000000 +0000
@@ -10,7 +10,7 @@
 [
   dnl We don't need (and can't compile) the replacement strtoll
   dnl unless the type 'long long' exists.
-  AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
   if test "$ac_cv_type_long_long" = yes; then
     AC_REPLACE_FUNCS(strtoll)
     if test $ac_cv_func_strtoll = no; then
diff -ur coreutils-5.2.1/m4/strtoull.m4 coreutils-5.2.1.new/m4/strtoull.m4
--- coreutils-5.2.1/m4/strtoull.m4	2003-08-17 07:51:49.000000000 +0000
+++ coreutils-5.2.1.new/m4/strtoull.m4	2005-04-16 20:09:07.000000000 +0000
@@ -10,7 +10,7 @@
 [
   dnl We don't need (and can't compile) the replacement strtoull
   dnl unless the type 'unsigned long long' exists.
-  AC_REQUIRE([jm_AC_TYPE_UNSIGNED_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_UNSIGNED_LONG_LONG])
   if test "$ac_cv_type_unsigned_long_long" = yes; then
     AC_REPLACE_FUNCS(strtoull)
     if test $ac_cv_func_strtoull = no; then
diff -ur coreutils-5.2.1/m4/strtoumax.m4 coreutils-5.2.1.new/m4/strtoumax.m4
--- coreutils-5.2.1/m4/strtoumax.m4	2003-09-13 10:15:46.000000000 +0000
+++ coreutils-5.2.1.new/m4/strtoumax.m4	2005-04-16 20:09:07.000000000 +0000
@@ -31,7 +31,7 @@
 
 # Prerequisites of lib/strtoumax.c.
 AC_DEFUN([gl_PREREQ_STRTOUMAX], [
-  jm_AC_TYPE_UINTMAX_T
+  gl_AC_TYPE_UINTMAX_T
   AC_CHECK_DECLS(strtoull)
-  AC_REQUIRE([jm_AC_TYPE_UNSIGNED_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_UNSIGNED_LONG_LONG])
 ])
diff -ur coreutils-5.2.1/m4/uintmax_t.m4 coreutils-5.2.1.new/m4/uintmax_t.m4
--- coreutils-5.2.1/m4/uintmax_t.m4	2003-08-10 08:47:53.000000000 +0000
+++ coreutils-5.2.1.new/m4/uintmax_t.m4	2005-04-16 20:09:07.000000000 +0000
@@ -13,12 +13,12 @@
 # Define uintmax_t to 'unsigned long' or 'unsigned long long'
 # if it is not already defined in <stdint.h> or <inttypes.h>.
 
-AC_DEFUN([jm_AC_TYPE_UINTMAX_T],
+AC_DEFUN([gl_AC_TYPE_UINTMAX_T],
 [
-  AC_REQUIRE([jm_AC_HEADER_INTTYPES_H])
-  AC_REQUIRE([jm_AC_HEADER_STDINT_H])
+  AC_REQUIRE([gl_AC_HEADER_INTTYPES_H])
+  AC_REQUIRE([gl_AC_HEADER_STDINT_H])
   if test $jm_ac_cv_header_inttypes_h = no && test $jm_ac_cv_header_stdint_h = no; then
-    AC_REQUIRE([jm_AC_TYPE_UNSIGNED_LONG_LONG])
+    AC_REQUIRE([gl_AC_TYPE_UNSIGNED_LONG_LONG])
     test $ac_cv_type_unsigned_long_long = yes \
       && ac_type='unsigned long long' \
       || ac_type='unsigned long'
diff -ur coreutils-5.2.1/m4/ulonglong.m4 coreutils-5.2.1.new/m4/ulonglong.m4
--- coreutils-5.2.1/m4/ulonglong.m4	2003-08-17 07:51:49.000000000 +0000
+++ coreutils-5.2.1.new/m4/ulonglong.m4	2005-04-16 20:09:07.000000000 +0000
@@ -10,7 +10,7 @@
 
 # Define HAVE_UNSIGNED_LONG_LONG if 'unsigned long long' works.
 
-AC_DEFUN([jm_AC_TYPE_UNSIGNED_LONG_LONG],
+AC_DEFUN([gl_AC_TYPE_UNSIGNED_LONG_LONG],
 [
   AC_CACHE_CHECK([for unsigned long long], ac_cv_type_unsigned_long_long,
   [AC_TRY_LINK([unsigned long long ull = 1ULL; int i = 63;],
diff -ur coreutils-5.2.1/m4/vasnprintf.m4 coreutils-5.2.1.new/m4/vasnprintf.m4
--- coreutils-5.2.1/m4/vasnprintf.m4	2004-01-16 08:26:02.000000000 +0000
+++ coreutils-5.2.1.new/m4/vasnprintf.m4	2005-04-16 20:09:07.000000000 +0000
@@ -24,7 +24,7 @@
 AC_DEFUN([gl_PREREQ_PRINTF_ARGS],
 [
   AC_REQUIRE([bh_C_SIGNED])
-  AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
   AC_REQUIRE([gt_TYPE_LONGDOUBLE])
   AC_REQUIRE([gt_TYPE_WCHAR_T])
   AC_REQUIRE([gt_TYPE_WINT_T])
@@ -33,7 +33,7 @@
 # Prequisites of lib/printf-parse.h, lib/printf-parse.c.
 AC_DEFUN([gl_PREREQ_PRINTF_PARSE],
 [
-  AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
   AC_REQUIRE([gt_TYPE_LONGDOUBLE])
   AC_REQUIRE([gt_TYPE_WCHAR_T])
   AC_REQUIRE([gt_TYPE_WINT_T])
@@ -46,7 +46,7 @@
 AC_DEFUN([gl_PREREQ_VASNPRINTF],
 [
   AC_REQUIRE([AC_FUNC_ALLOCA])
-  AC_REQUIRE([jm_AC_TYPE_LONG_LONG])
+  AC_REQUIRE([gl_AC_TYPE_LONG_LONG])
   AC_REQUIRE([gt_TYPE_LONGDOUBLE])
   AC_REQUIRE([gt_TYPE_WCHAR_T])
   AC_REQUIRE([gt_TYPE_WINT_T])
diff -ur coreutils-5.2.1/m4/xstrtol.m4 coreutils-5.2.1.new/m4/xstrtol.m4
--- coreutils-5.2.1/m4/xstrtol.m4	2003-09-13 10:45:30.000000000 +0000
+++ coreutils-5.2.1.new/m4/xstrtol.m4	2005-04-16 20:09:07.000000000 +0000
@@ -16,7 +16,7 @@
 AC_DEFUN([gl_PREREQ_XSTRTOL_H],
 [
   AC_REQUIRE([jm_AC_TYPE_INTMAX_T])
-  AC_REQUIRE([jm_AC_TYPE_UINTMAX_T])
+  AC_REQUIRE([gl_AC_TYPE_UINTMAX_T])
 ])
 
 # Prerequisites of lib/xstrtol.c.
diff -ur coreutils-5.2.1/m4/xstrtoumax.m4 coreutils-5.2.1.new/m4/xstrtoumax.m4
--- coreutils-5.2.1/m4/xstrtoumax.m4	2003-08-17 07:51:49.000000000 +0000
+++ coreutils-5.2.1.new/m4/xstrtoumax.m4	2005-04-16 20:09:07.000000000 +0000
@@ -4,6 +4,6 @@
 AC_DEFUN([jm_XSTRTOUMAX],
 [
   dnl Prerequisites of lib/xstrtoumax.c.
-  AC_REQUIRE([jm_AC_TYPE_UINTMAX_T])
+  AC_REQUIRE([gl_AC_TYPE_UINTMAX_T])
   AC_REQUIRE([gl_PREREQ_XSTRTOL])
 ])
