diff -urN coreutils-6.7.org/m4/openat.m4 coreutils-6.7/m4/openat.m4
--- coreutils-6.7.org/m4/openat.m4	2006-11-21 00:32:38.000000000 +0100
+++ coreutils-6.7/m4/openat.m4	2006-12-12 00:02:35.949793500 +0100
@@ -10,6 +10,12 @@
 
 AC_DEFUN([gl_FUNC_OPENAT],
 [
+ needcompat=no
+ AC_CHECK_FUNC(mkdirat, [], [needcompat=yes])
+ AC_CHECK_FUNC(fchmodat, [], [needcompat=yes])
+ AC_CHECK_FUNC(fchownat, [], [needcompat=yes])
+
+ if test x$needcompat = xyes; then
   # No system provides these functions; compile them unconditionally.
   AC_LIBOBJ([mkdirat])
   AC_LIBOBJ([fchmodat])
@@ -28,6 +34,7 @@
       [Define to rpl_ if the openat replacement function should be used.])
     gl_PREREQ_OPENAT;;
   esac
+ fi
 ])
 
 AC_DEFUN([gl_PREREQ_OPENAT],
