--- coreutils-5.0/src/head.c.orig	Tue Oct  1 10:47:44 2002
+++ coreutils-5.0/src/head.c	Wed Jan 14 14:54:24 2004
@@ -343,7 +343,7 @@
 	    }
 	}
 
-      if (200112 <= posix2_version ())
+      if ((200112 <= posix2_version ()) && getenv("POSIXLY_CORRECT"))
 	{
 	  error (0, 0, _("`-%s' option is obsolete; use `-%c %.*s%.*s%s'"),
 		 n_string, count_lines ? 'n' : 'c',
--- coreutils-5.0/src/tail.c.orig	Sat Dec 14 18:26:21 2002
+++ coreutils-5.0/src/tail.c	Wed Jan 14 14:55:59 2004
@@ -1298,7 +1298,7 @@
   if (argc < 2)
     return 0;
 
-  obsolete_usage = (posix2_version () < 200112);
+  obsolete_usage = ((posix2_version () < 200112) || !getenv("POSIXLY_CORRECT"));
 
   /* If P starts with `+' and the POSIX version predates 1003.1-2001,
      or if P starts with `-N' (where N is a digit), or `-l', then it
